Let PartitionedDataSet lazily materialize data to save

lou-k opened the pull request:
## Description
The `PartititionedDataSet` is great for very large data sets (i.e., ones that won't fit in RAM), but it's `save` function requires all data frames to be in memory. As discussed in https://github.com/quantumblacklabs/kedro/issues/499, it would be advantageous for the `save` function to support a dictionary of callables, much like the `load` function returns. This would enable an entire large dataset to be processed and moved down a pipeline without storing the whole thing in RAM.

## Development notes
* Added a check in `PartititionedDataSet.save` to see if the value of the dictionary is a callable. If it is, it calls it.
* Add a unit test for `PartititionedDataSet.save` for this case
* Updated the documentation page for this case.
* As noted in the [issue](https://github.com/quantumblacklabs/kedro/issues/499), this causes saves to occur after the node_run hook, which may cause unexpected behavior. I've noted this in the updated documentation.

## Checklist

- [x] Read the [contributing](https://github.com/quantumblacklabs/kedro/blob/master/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change and added my name to the list of supporting contributions in the [`RELEASE.md`](https://github.com/quantumblacklabs/kedro/blob/master/RELEASE.md) file
- [x] Added tests to cover my changes

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

markjunod commented:
This is an amazing addition!  I've run into this issue many times since I started using Kedro.

deepyaman commented on RELEASE.md:
```diff
@@ -12,10 +12,12 @@
 * CLI commands from sources with the same name will show under one list in the help screen.
 * The setup of a Kedro project, including adding src to path and configuring settings, is now handled via the `bootstrap_project` method.
 * Invoked `configure_project` if a `package_name` is supplied to `KedroSession.create`. This is added for backward-compatibility purpose to support workflow that creates a `Session` manually. It will only be removed in `0.18.0`.
+* `PartitionedDataSet` now supports lazily materilizing data on save.
```
```suggestion
* `PartitionedDataSet` now supports lazily materializing data on save.
```

lou-k replied:
Thanks!

deepyaman commented on RELEASE.md:
```diff
@@ -12,10 +12,12 @@
 ## Minor breaking changes to the API
 
 ## Thanks for supporting contributions
+[Lou Kratz](https://github.com/lou-k),
```
```suggestion
[Lou Kratz](https://github.com/lou-k)
```

deepyaman commented on docs/source/05_data/02_kedro_io.md:
```diff
@@ -466,6 +466,29 @@ def create_partitions() -> Dict[str, Any]:
 
 > *Note:* Writing to an existing partition may result in its data being overwritten, if this case is not specifically handled by the underlying dataset implementation. You should implement your own checks to ensure that no existing data is lost when writing to a `PartitionedDataSet`. The simplest safety mechanism could be to use *partition IDs* that have a high chance of uniqueness - for example, ...
 
+`PartitionedDataSet` also supports lazy saving, where the partition's data is not materialized until the it's time to write.
```
```suggestion
`PartitionedDataSet` also supports lazy saving, where the partition's data is not materialized until it's time to write.
```

deepyaman commented on docs/source/05_data/02_kedro_io.md:
```diff
@@ -466,6 +466,29 @@ def create_partitions() -> Dict[str, Any]:
+    return {
+        # create a file "s3://my-bucket-name/part/foo.csv"
+        "part/foo": lambda: pd.DataFrame({"data": [1, 2]}),
+        # create a file "s3://my-bucket-name/part/bar.csv.csv"
```
```suggestion
        # create a file "s3://my-bucket-name/part/bar.csv"
```

deepyaman reviewed (commented):
Noted a few minor edits as I read through.

lou-k commented:
Accepted all of @deepyaman 's suggested and rebased/squashed.

lucasjamar commented:
Hi @lou-k ,

Super cool contribution! Just out of curiosity, from your experience in implementing this solution so far, do you think it would be quite easy or rather difficult to extend this property to IncrementalDataSet?

lou-k commented:
@lucasjamar since `IncrementalDataSet` inherits from `PartitionedDataSet` you'd get this for free there.

MerelTheisenQB commented on docs/source/05_data/02_kedro_io.md:
````diff
@@ -459,13 +459,36 @@ def create_partitions() -> Dict[str, Any]:
+    }
+```
+
+> *Note:* Lazy saving causes the dataset to be written _after_ the `after_node_run` [hook](../07_extend_kedro/02_hooks), which could cause unintended behavior. 
````
```suggestion
> *Note:* Lazy saving causes the dataset to be written _after_ the `after_node_run` [hook](../07_extend_kedro/02_hooks), which could cause unintended behaviour. 
```

lou-k replied:
Whoops 😅 . I've followed your recommendations for changing this sentence below.

MerelTheisenQB commented on docs/source/05_data/02_kedro_io.md:
````diff
@@ -459,13 +459,36 @@ def create_partitions() -> Dict[str, Any]:
+    }
+```
+
+> *Note:* Lazy saving causes the dataset to be written _after_ the `after_node_run` [hook](../07_extend_kedro/02_hooks), which could cause unintended behavior. 
````
Is there anything you could do to avoid this potential unintended behaviour? If not, perhaps it's better to just rephrase this to: *Note:* When using lazy saving the dataset will be written _after_ the `after_node_run` [hook](../07_extend_kedro/02_hooks).

Otherwise it kind of sounds like there is maybe something you could do differently.

lou-k replied:
I don't believe there is a way to avoid this (I took this note from the original issues discussion). I'll update the documentation to make this clearer. Thanks so much!

MerelTheisenQB commented on RELEASE.md:
```diff
@@ -12,10 +12,12 @@
 * CLI commands from sources with the same name will show under one list in the help screen.
 * The setup of a Kedro project, including adding src to path and configuring settings, is now handled via the `bootstrap_project` method.
 * Invoked `configure_project` if a `package_name` is supplied to `KedroSession.create`. This is added for backward-compatibility purpose to support workflow that creates a `Session` manually. It will only be removed in `0.18.0`.
+* `PartitionedDataSet` now supports lazily materializing data on save.
 
 ## Minor breaking changes to the API
 
 ## Thanks for supporting contributions
+[Lou Kratz](https://github.com/lou-k)
```
Could you move this to the section for `0.17.4`?

lou-k replied:
👍

MerelTheisenQB reviewed (approved):
Thanks for this contribution @lou-k ! And also for your patience with the review. I left some minor comments, but otherwise this looks good to be merged in 👍

lou-k commented:
Thanks so much for the review @MerelTheisenQB !

I've:
* Updated the documentation with your suggestions. Thanks for that!
* Rebased against the latest master branch
* Moved the contributions to `0.17.4`
* Squashed the commits for a clean merge.

🤞 the CI/CD passes again :D

lou-k commented:
Hmm I am unsure why the `win_unit_tests_38 ` tests are failing. I ran `make test` on my macos machine. Can anyone share the circleci logs?

MerelTheisenQB commented:
> Hmm I am unsure why the `win_unit_tests_38 ` tests are failing. I ran `make test` on my macos machine. Can anyone share the circleci logs?

Unfortunately, these tests are very flaky and we haven't yet managed to fix them properly 😢  They are not required to pass the build at the moment though. You just need a second approval and then the PR can be merged.

ignacioparicio reviewed (approved):
LGTM, thank you @lou-k for the contribution!

banbiossa commented:
Just what I need but available only on 0.17.4 I guess 😢

MerelTheisenQB commented:
Yes @banbiossa it's available from Kedro 0.17.4, which was released last week 🙂

banbiossa commented:
Not totally related but lambda scopes bit me on using this 😿 
https://stackoverflow.com/questions/938429/scope-of-lambda-functions-and-their-parameters

lou-k commented:
@banbiossa SAME. EVERY. TIME.

yurigba commented:
I am using this implementation for a image dataset that won't fit in my memory, yet I still am having memory issues when saving the dataset. Any tips on how can I debug this?
